stan/math/opencl/matrix_cl.hpp

```diff
@@ -416,63 +418,46 @@ class matrix_cl<T, enable_if_arithmetic<T>> {
   /**
    * Assign a \c matrix_cl to another
    */
-  matrix_cl<T>& operator=(const matrix_cl<T>& a) {
-    if (a.size() == 0) {
-      return *this;
-    }
+  matrix_cl<T>& operator=(matrix_cl<T>&& a) {
     view_ = a.view();
     rows_ = a.rows();
     cols_ = a.cols();
     this->wait_for_read_write_events();
-    cl::CommandQueue queue = opencl_context.queue();
-    try {
-      cl::Event copy_event;
-      queue.enqueueCopyBuffer(a.buffer(), this->buffer(), 0, 0,
-                              a.size() * sizeof(T), &a.write_events(),
-                              &copy_event);
-      this->add_write_event(copy_event);
-      a.add_read_event(copy_event);
-    } catch (const cl::Error& e) {
-      check_opencl_error("copy (OpenCL)->(OpenCL)", e);
-    }
-    return *this;
-  }
-
-  /**
-   * Move a \c matrix_cl to another
-   */
-  matrix_cl<T>& operator=(matrix_cl<T>&& a) {
-    // Need to wait for all of matrices events before destroying old buffer
-    this->wait_for_read_write_events();
-    buffer_cl_ = a.buffer();
-    view_ = a.view();
+    buffer_cl_ = std::move(a.buffer_cl_);
     write_events_ = std::move(a.write_events_);
     read_events_ = std::move(a.read_events_);
-    rows_ = a.rows();
-    cols_ = a.cols();
     return *this;
   }
 
   /**
-   * Assign a \c matrix_cl of one arithmetic type to another
+   * Assign a \c matrix_cl to another
    */
-  template <typename U, typename = enable_if_arithmetic<U>>
-  matrix_cl<T>& operator=(const matrix_cl<U>& a) {
-    // Need to wait for all of matrices events before destroying old buffer
-    this->wait_for_read_write_events();
-    buffer_cl_ = a.buffer();
-    view_ = a.view();
-    rows_ = a.rows();
-    cols_ = a.cols();
+  matrix_cl<T>& operator=(const matrix_cl<T>& a) {
+    this->view_ = a.view();
+    this->rows_ = a.rows();
+    this->cols_ = a.cols();
+    cl::Context& ctx = opencl_context.context();
+    cl::CommandQueue queue = opencl_context.queue();
+    try {
```

Comment:
Why did you remove waiting for read and write events on this? I think we kinda need it.

Reply:
I think we only need to wait for the write events in the copyBuffer function? Cuz reads in this context are fine, its just the writes we worry about

```cpp
      queue.enqueueCopyBuffer(a.buffer(), this->buffer(), 0, 0,
                              a.size() * sizeof(T), &a.write_events(),
                              &cstr_event);
```